presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java

```diff
@@ -149,10 +150,8 @@
                 newSourceBuilder.build(),
                 inputsBuilder.build());
 
-        if (!result.getOutputSymbols().equals(project.getOutputSymbols())) {
-            // we need to strip unnecessary symbols (hash, partitioning columns).
-            result = new ProjectNode(idAllocator.getNextId(), result, identity(project.getOutputSymbols()));
-        }
+        // we need to strip unnecessary symbols (hash, partitioning columns).
+        result = restrictOutputs(idAllocator, result, ImmutableSet.copyOf(project.getOutputSymbols()));
```

Comment:
If project had a larger set of output symbols than result, then the previous code would have produced an invalid identity projection that referred to non-existing symbols.